File: packages/sdk/dart/packages/flutter/lib/src/auth_refresh.dart
import 'dart:convert';

import 'package:edgebase_core/src/errors.dart';
import 'package:edgebase_core/src/token_manager.dart' as core;
import 'package:http/http.dart' as http;

String? _extractServerMessage(String? rawBody) {
  if (rawBody == null || rawBody.isEmpty) return null;
  try {
    final decoded = jsonDecode(rawBody);
    if (decoded is Map<String, dynamic>) {
      for (final key in ['message', 'error', 'detail']) {
        final value = decoded[key];
        if (value is String && value.trim().isNotEmpty) {
          return value.trim();
        }
      }
    }
  } catch (_) {
    // Ignore malformed response bodies and fall back to a synthesized message.
  }
  return null;
}

Future<core.TokenPair> refreshAccessToken(
  String baseUrl,
  String refreshToken,
) async {
  final refreshUrl = '${baseUrl.replaceAll(RegExp(r'/$'), '')}/api/auth/refresh';
  late http.Response response;

  try {
    response = await http.post(
      Uri.parse(refreshUrl),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode({'refreshToken': refreshToken}),
    );
  } catch (error) {
    throw EdgeBaseError(
      'Auth session refresh could not reach $refreshUrl. Make sure the EdgeBase server is running and reachable. Cause: $error',
      statusCode: 0,
    );
  }

  final body = response.body.isEmpty
      ? null
      : jsonDecode(response.body) as Map<String, dynamic>;
  if (response.statusCode < 200 || response.statusCode >= 300) {
    throw EdgeBaseError(
      body?['message'] as String? ??
          _extractServerMessage(response.body) ??
          'Request failed with HTTP ${response.statusCode} and no error message from the server.',
      statusCode: response.statusCode,
    );
  }

  final accessToken = body?['accessToken'] as String?;
  final nextRefreshToken = body?['refreshToken'] as String?;
  if (accessToken == null || nextRefreshToken == null) {
    throw EdgeBaseError(
      'Invalid auth refresh response.',
      statusCode: response.statusCode,
    );
  }

  return core.TokenPair(
    accessToken: accessToken,
    refreshToken: nextRefreshToken,
  );
}

File: packages/sdk/js/packages/web/src/auth-refresh.ts
import { EdgeBaseError, networkError, parseErrorResponse } from '@edge-base/core';

interface RefreshResponse {
  accessToken?: string;
  refreshToken?: string;
  message?: string;
}

const COOKIE_REFRESH_TIMEOUT_MS = 15_000;

export async function refreshAccessToken(
  baseUrl: string,
  refreshToken: string,
  transport: 'body' | 'httpOnlyCookie' = 'body',
): Promise<{
  accessToken: string;
  refreshToken: string;
}> {
  const refreshUrl = `${baseUrl.replace(/\/$/, '')}/api/auth/refresh`;
  const useCookie = transport === 'httpOnlyCookie';
  let response: Response;
  let body: RefreshResponse | null;
  const controller = useCookie ? new AbortController() : null;
  let timedOut = false;
  let timeout: ReturnType<typeof setTimeout> | undefined;

  try {
    const request = (async () => {
      const nextResponse = await fetch(refreshUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...(useCookie ? { 'X-EdgeBase-Auth-Transport': 'cookie' } : {}),
        },
        ...(useCookie ? { credentials: 'include' as const } : {}),
        body: JSON.stringify(refreshToken ? { refreshToken } : {}),
        ...(controller ? { signal: controller.signal } : {}),
      });
      const nextBody = await nextResponse.json().catch(() => null) as RefreshResponse | null;
      return { response: nextResponse, body: nextBody };
    })();
    const result = useCookie
      ? await Promise.race([
          request,
          new Promise<never>((_resolve, reject) => {
            timeout = setTimeout(() => {
              timedOut = true;
              controller?.abort();
              reject(new Error(`Cookie auth refresh timed out after ${COOKIE_REFRESH_TIMEOUT_MS}ms.`));
            }, COOKIE_REFRESH_TIMEOUT_MS);
          }),
        ])
      : await request;
    response = result.response;
    body = result.body;
  } catch (error) {
    throw networkError(
      timedOut
        ? `Auth session refresh timed out while reaching ${refreshUrl}.`
        : `Auth session refresh could not reach ${refreshUrl}. Make sure the EdgeBase server is running and reachable.`,
      { cause: error },
    );
  } finally {
    if (timeout !== undefined) clearTimeout(timeout);
  }

  if (!response.ok) {
    throw parseErrorResponse(response.status, body);
  }

  if (!body?.accessToken || (!useCookie && !body?.refreshToken)) {
    throw new EdgeBaseError(
      500,
      useCookie
        ? 'Auth refresh succeeded but did not return an accessToken. Check the server auth configuration.'
        : 'Auth refresh succeeded but did not return both accessToken and refreshToken. Check the server auth configuration.',
    );
  }

  return {
    accessToken: body.accessToken,
    refreshToken: body.refreshToken ?? '',
  };
}
